telegram/ext/_picklepersistence.py

```diff
@@ -26,14 +28,99 @@
+
+
+def _all_subclasses(cls: Type[TelegramObject]) -> Set[Type[TelegramObject]]:
+    """Gets all subclasses of the specified object, recursively. from
+    https://stackoverflow.com/a/3862957/9706202
+    """
+    subclasses = cls.__subclasses__()
+    return set(subclasses).union([s for c in subclasses for s in _all_subclasses(c)])
+
+
+def _reconstruct_to(cls: Type[TelegramObject], kwargs: dict) -> TelegramObject:
+    """
+    This method is used for unpickling. The data, which is in the form a dictionary, is
+    converted back into a class. Functions the same as :meth:`TelegramObject.__setstate__`
+    """
+    obj = cls(**kwargs)
+    if '_bot' in kwargs:
+        obj.set_bot(kwargs['_bot'])
+    return obj
+
+
+def _custom_reduction(cls: TelegramObject) -> Tuple[Callable, Tuple[Type[TelegramObject], dict]]:
+    """
+    This method is used for pickling. The bot attribute is preserved so _BotPickler().persistent_id
+    works as intended.
+    """
+    # Adapted from self.to_dict(), unlike to_dict(), this is not recursive.
+    data = {}
+    attrs = {attr for cls in cls.__class__.__mro__[:-1] for attr in cls.__slots__}
+    for key in attrs:
+        value = getattr(cls, key, None)
+        if value is not None:
+            data[key] = value
+
+    return _reconstruct_to, (cls.__class__, data)
+
+
+class _BotPickler(pickle.Pickler):
+    __slots__ = ('_bot',)
+
+    # Here we define a private dispatch_table, because we want to preserve the bot attribute of
+    # objects so persistent_id works as intended. Otherwise, the bot attribute is deleted in
+    # __getstate__, which used for regular pickling (via pickle.dumps(...))
+    dispatch_table = copyreg.dispatch_table.copy()  # type: ignore[attr-defined]
+    for obj in _all_subclasses(TelegramObject):
+        dispatch_table[obj] = _custom_reduction
+
+    def __init__(self, bot: Bot, *args: Any, **kwargs: Any):
+        self._bot = bot
+        super().__init__(*args, **kwargs)
+
+    def persistent_id(self, obj: object) -> Optional[str]:
+        """Used to 'mark' the Bot, so it can be replaced later. See
+        https://docs.python.org/3/library/pickle.html#pickle.Pickler.persistent_id for more info
+        """
+        if obj is self._bot:
+            return _REPLACED_KNOWN_BOT
+        if isinstance(obj, Bot):
+            warn(
+                'Unknown bot instance found. Will be replaced by `None` during unpickling',
+                stacklevel=7,
+            )
+            return _REPLACED_UNKNOWN_BOT
+        return None  # pickles as usual
+
+
+class _BotUnpickler(pickle.Unpickler):
+    __slots__ = ('_bot',)
+
+    def __init__(self, bot: Bot, *args: Any, **kwargs: Any):
+        self._bot = bot
+        super().__init__(*args, **kwargs)
+
+    def persistent_load(self, pid: str) -> Optional[Bot]:
+        """Replaces the bot with the current bot if known, else it is replaced by :obj:`None`."""
+        return self._bot if pid == _REPLACED_KNOWN_BOT else None
```

Comment:
to increase backwards compatibility, we could adjust this check such that it also accepts the current constant `BasePersistence.REPLACED_BOT` (or whatever the current name is). We could even just use the current value for `_R_K_B`.
It would also be nice to have a unit test for the backwards compatibility - one could dump some test data with the current implementation, save it in `tests/data` and make sure that it's correctly unpickled with the new implementation

Reply:
Good idea!

Reply:
On second thought, this would not be possible, since `persistent_load` will only run when the data has been tagged by the `persistent_id` during pickling. So for backwards compatibility, in the automatic transition script, we should:

1. Subclass `Bot` to add back [`__reduce__`](https://github.com/python-telegram-bot/python-telegram-bot/pull/1994/files#diff-ca0526e3005b92a3844ee53d68688fe5a8c438332365a0efc3275a814cca5008L3814-L3816)
2. Ask user for their token, etc to make a `Bot`
3.  Use `BP.insert_bot` to find `BP.REPLACED_BOT` and replace with the bot in step 1
4. Pickle the data back, using `pickle.dumps`

But one problem I see with this is that when the user starts running their bot in v14, `persistent_id` will mark this as an unknown bot. To mitigate this, we could hardcode it so that it recognizes that this bot was for the transition only.

What do you think?

Reply:
Alternatively, we could add logic to `persistent_id` to handle `BP.REPLACED_BOT`, but the actual bot would be inserted only in the second bot restart...

Reply:
Ah, right. Mh, if we can't provide a proper backwards compatibility, then I guess a transition script is more sane. For that I guess it would suffice the load the data with very code and dump it with v14 code, no?

```python
from v13_pp import PP as PPv13
from tg.ext import PP as PPv14

ppv13 = PPv13(...)
ppv13.set_bot(...)
ppv14 = PPv14(...)
ppv14.set_bot(...)
chat_data = ppv13.get_chat_data()
pp14.update_chat_data(...)
pp14.flush()

```

Reply:
Yes.. that would work too I supoose